search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,603 @@
+                                                        NULL),
+                                         NULL,
+                                         G_DBUS_CALL_FLAGS_NONE,
+                                         -1,
+                                         NULL,
+                                         &error);
+
+    if (reply)
+        g_variant_unref (reply);
+    if (error) {
+        g_dbus_method_invocation_return_gerror (invocation, error);
+        g_error_free (error);
+        return;
+    }
+    g_dbus_method_invocation_return_value (invocation, NULL);
+    return;
+}
+
+static gboolean 
+handle_launch_search (YelpShellSearchProvider2  *skeleton,
+                      GDBusMethodInvocation     *invocation,
+                      gchar                    **terms,
+                      guint32                    timestamp,
+                      gpointer                   user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    gchar *joined_terms = g_strjoinv (" ", terms);
+
+    yelp_launch (joined_terms, app, invocation, "show-search");
+
+    g_free (joined_terms);
+
+    return TRUE;
+}
+
+static gboolean
+handle_activate_result (YelpShellSearchProvider2  *skeleton,
+                        GDBusMethodInvocation     *invocation,
+                        gchar                     *result,
+                        gchar                    **terms,
+                        guint32                    timestamp,
+                        gpointer                   user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+
+    yelp_launch (result, app, invocation, "show-page");
+
+    return TRUE;
+}
+
+static void
+search_provider_app_dispose (GObject *obj)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (obj);
+
+    g_clear_object (&self->skeleton);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
+
+    G_OBJECT_CLASS (yelp_search_provider_app_parent_class)->dispose (obj);
+}
+
+static void
+preload_data_cb (YelpDocument          *document,
+                 YelpDocumentSignal     signal,
+                 YelpSearchProviderApp *app,
+                 GError                *error)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_ERROR) {
+        g_warning ("error during preloading data: %s", (error != NULL) ? error->message : "unknown error");
+        return;
+    }
+
+    if (signal == YELP_DOCUMENT_SIGNAL_CONTENTS)
+        return;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_INFO) {
+        for (iter = page_ids; *iter; iter++) {
+            gchar *page_id = *iter;
+            PageData *data = g_new0 (PageData, 1);
+            gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+            GIcon *icon = g_icon_new_for_string (icon_string, NULL);
```

Comment:
I don't think this is right; you probably want to use `g_themed_icon_new()`, since `yelp_document_get_page_icon()` gets you icon name, not a serialized `GIcon`